File: backend/app/api/v1/templates.py
"""Endpoints de Plantillas de Documento y Extracción Preliminar."""
import uuid

from fastapi import APIRouter, File, Form, UploadFile, status

from app.api.deps import CurrentUserAny, DbSession
from app.schemas.template import ExtractedFieldPreview, TemplateCreate, TemplateRead, TemplateUpdate
from app.services.ocr.pipeline import run_ocr
from app.services.ocr.pdf_rasterizer import is_pdf
from app.services.template_service import TemplateService

router = APIRouter(prefix="/templates", tags=["templates"])


@router.get("", response_model=list[TemplateRead])
def list_templates(user: CurrentUserAny, db: DbSession):
    return TemplateService(db).list_templates(user)


@router.post("", response_model=TemplateRead, status_code=status.HTTP_201_CREATED)
def create_template(payload: TemplateCreate, user: CurrentUserAny, db: DbSession):
    return TemplateService(db).create_template(
        user, payload.name, payload.description, [f.model_dump() for f in payload.field_definitions]
    )


@router.get("/{template_id}", response_model=TemplateRead)
def get_template(template_id: uuid.UUID, user: CurrentUserAny, db: DbSession):
    return TemplateService(db).get_template(user, template_id)


@router.put("/{template_id}", response_model=TemplateRead)
def update_template(template_id: uuid.UUID, payload: TemplateUpdate, user: CurrentUserAny, db: DbSession):
    return TemplateService(db).update_template(
        user, template_id, payload.name, payload.description,
        [f.model_dump() for f in payload.field_definitions] if payload.field_definitions else None,
    )


@router.delete("/{template_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_template(template_id: uuid.UUID, user: CurrentUserAny, db: DbSession):
    TemplateService(db).delete_template(user, template_id)


@router.post("/preview-extraction", response_model=list[ExtractedFieldPreview])
def preview_extraction(user: CurrentUserAny, db: DbSession, file: UploadFile = File(...)):
    """Sube un documento, ejecuta OCR + LLM y devuelve todos los campos detectados para que el usuario seleccione cuáles guardar en la plantilla."""
    content = file.file.read()
    ocr_text = run_ocr(content)
    raw_fields = TemplateService(db).preview_extraction(ocr_text)
    return [
        ExtractedFieldPreview(
            name=f.get("name", "campo_desconocido"),
            label=f.get("label", f.get("name", "Campo")),
            value=f.get("value"),
            confidence=f.get("confidence", 0),
        )
        for f in raw_fields
    ]


@router.post("/from-result", response_model=TemplateRead, status_code=status.HTTP_201_CREATED)
def create_template_from_result(payload: TemplateCreate, user: CurrentUserAny, db: DbSession):
    """Crea una plantilla a partir de los campos de un resultado de extracción ya procesado."""
    return TemplateService(db).create_template(
        user, payload.name, payload.description, [f.model_dump() for f in payload.field_definitions]
    )

File: frontend/src/api/templates.ts
import { apiClient } from "./client";

export interface FieldDefinition {
  name: string;
  label: string;
}

export interface TemplateRead {
  id: string;
  user_id: string | null;
  name: string;
  description: string;
  field_definitions: FieldDefinition[];
  is_active: boolean;
  is_draft: boolean;
  created_at: string;
}

export interface ExtractedFieldPreview {
  name: string;
  label: string;
  value: string | number | null;
  confidence: number;
}

export async function listTemplates(): Promise<TemplateRead[]> {
  const { data } = await apiClient.get("/templates");
  return data;
}

export async function getTemplate(id: string): Promise<TemplateRead> {
  const { data } = await apiClient.get(`/templates/${id}`);
  return data;
}

export async function createTemplate(payload: {
  name: string;
  description: string;
  field_definitions: FieldDefinition[];
  make_global?: boolean;
}): Promise<TemplateRead> {
  const { data } = await apiClient.post("/templates", payload);
  return data;
}

export async function updateTemplate(
  id: string,
  payload: { name?: string; description?: string; field_definitions?: FieldDefinition[] }
): Promise<TemplateRead> {
  const { data } = await apiClient.put(`/templates/${id}`, payload);
  return data;
}

export async function deleteTemplate(id: string): Promise<void> {
  await apiClient.delete(`/templates/${id}`);
}

export async function previewExtraction(file: File): Promise<ExtractedFieldPreview[]> {
  const formData = new FormData();
  formData.append("file", file);
  const { data } = await apiClient.post("/templates/preview-extraction", formData, {
    headers: { "Content-Type": "multipart/form-data" },
  });
  return data;
}

// --- Administración de plantillas (HU 4.2) ---

export async function listDraftTemplates(): Promise<TemplateRead[]> {
  const { data } = await apiClient.get("/admin/templates/drafts");
  return data;
}

export async function approveDraftTemplate(
  id: string,
  payload: { name?: string; description?: string; field_definitions?: FieldDefinition[]; make_global?: boolean }
): Promise<TemplateRead> {
  const { data } = await apiClient.post(`/admin/templates/drafts/${id}/approve`, payload);
  return data;
}

export async function rejectDraftTemplate(id: string): Promise<void> {
  await apiClient.delete(`/admin/templates/drafts/${id}`);
}

export async function setTemplateActive(id: string, is_active: boolean): Promise<TemplateRead> {
  const { data } = await apiClient.patch(`/admin/templates/${id}/active`, { is_active });
  return data;
}

export async function promoteTemplateToGlobal(id: string): Promise<TemplateRead> {
  const { data } = await apiClient.patch(`/admin/templates/${id}/promote-global`, {});
  return data;
}
